shell/golangci-lint.sh

```diff
@@ -17,8 +17,45 @@ fi
 # Enable only fast linters, and always use the correct config.
 args=("--config=${workspaceFolder}/scripts/golangci.yml" "$@" "--fast" "--allow-parallel-runners")
 
-# trade memory for CPU when running the linters
-export GOGC=20
+# Determine the version of go and golangci-lint to calculate compatibility.
+GO_MINOR_VERSION=$(go version | awk '{print $3}' | sed 's/go//' | cut -d'.' -f1,2)
+GOLANGCILINT_VERSION=$(asdf_devbase_run golangci-lint --version | awk '{print $4}')
+GO_MINOR_VERSION_INT=${GO_MINOR_VERSION//./}
+GOLANGCI_LINT_VERSION_INT=${GOLANGCILINT_VERSION//./}
+
+# Go 1.20 requires >= golangci-lint 1.52.0
+if [[ $GO_MINOR_VERSION_INT == 120 ]] && [[ $GOLANGCI_LINT_VERSION_INT -lt 1520 ]]; then
+  echo "Error: Go 1.20 requires golangci-lint 1.52.0 or newer (detected $GOLANGCILINT_VERSION)" >&2
+  exit 1
+fi
+
+# Go 1.21 requires >= golangci-lint 1.54.1
+if [[ $GO_MINOR_VERSION_INT == 121 ]] && [[ $GOLANGCI_LINT_VERSION_INT -lt 1541 ]]; then
+  echo "Error: Go 1.21 requires golangci-lint 1.54.1 or newer (detected $GOLANGCILINT_VERSION)" >&2
+  exit 1
+fi
+
+# If we're on a system with free, set GOMEMLIMIT to a value that's less
+# than the max amount of RAM on the system. This helps ensure that we
+# don't go over the memory limit and get OOM killed. This is mostly
+# important for CI systems.
+if (command -v free || command -v sysctl) &>/dev/null; then
+  if command -v free &>/dev/null; then
+    mem="$(free -m | awk '/^Mem:/{print $2}')"
+  elif command -v sysctl &>/dev/null; then
+    mem=$((($(sysctl -n hw.memsize) / 1024) / 1024))
+  fi
+
+  if [[ $mem -lt 1024 ]] || [[ -z $mem ]]; then
```

Comment:
suggestion: make it easier to tweak the magic number
```suggestion
  RESERVED_MEMORY_IN_MIB=1024
  if [[ $mem -lt $RESERVED_MEMORY_IN_MIB ]] || [[ -z $mem ]]; then
```

Reply:
I was thinking about that. Good call :D